File: 2/2.js
const fs = require("fs");

function getData(file) {
  let rawData = fs.readFileSync(file, { encoding: "utf-8" });
  let data = rawData.trim().split("\n");
  data = data.map((el) => {
    tuple = el.split(" ");
    return [tuple[0].toLowerCase(), parseInt(tuple[1])];
  });
  return data;
}

function calculatePosition(data) {
  let [x, depth] = [0, 0];
  for (const tuple of data) {
    value = tuple[1];
    switch (tuple[0]) {
      case "forward":
        x += value;
        break;
      case "down":
        depth += value;
        break;
      case "up":
        depth -= value;
        break;
      default:
        break;
    }
  }
  return x * depth;
}

function calculatePositionWithAim(data) {
  let [x, depth, aim] = [0, 0, 0];
  for (const tuple of data) {
    value = tuple[1];
    switch (tuple[0]) {
      case "forward":
        x += value;
        depth += aim * value;
        break;
      case "down":
        // depth += value;
        aim += value;
        break;
      case "up":
        // depth -= value;
        aim -= value;
        break;
      default:
        break;
    }
  }
  return x * depth;
}

function main() {
  const data = getData("input.txt");
  const res1 = calculatePosition(data);
  const res2 = calculatePositionWithAim(data);
  console.log(`Part 1: ${res1}`);
  console.log(`Part 2: ${res2}`);
}

main();

File: 2/2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type movement struct {
	direction string
	value     int
}

func getData(fileName string) []movement {
	rawByteData, err := ioutil.ReadFile(fileName)
	check(err)
	rawData := strings.TrimSpace(string(rawByteData))
	rawDataArray := strings.Split(rawData, "\n")
	data := make([]movement, len(rawDataArray))
	for i, v := range rawDataArray {
		tuple := strings.Split(v, " ")
		numericValue, err := strconv.Atoi(tuple[1])
		check(err)
		data[i] = movement{direction: tuple[0], value: numericValue}
	}
	return data
}

func calculatePosition(data []movement) int {
	var x, depth int
	for _, mov := range data {
		movValue := mov.value
		switch mov.direction {
		case "forward":
			x += movValue
		case "down":
			depth += movValue
		case "up":
			depth -= movValue
		}
	}
	return x * depth
}

func calculatePositionWithAim(data []movement) int {
	var x, depth, aim int
	for _, mov := range data {
		movValue := mov.value
		switch mov.direction {
		case "forward":
			x += movValue
			depth += aim * movValue
		case "down":
			aim += movValue
		case "up":
			aim -= movValue
		}
	}
	return x * depth
}

func main() {
	data := getData("input.txt")
	res1 := calculatePosition(data)
	res2 := calculatePositionWithAim(data)

	fmt.Println("Part 1: ", res1)
	fmt.Println("Part 2: ", res2)
}
